### DISCUSSION TITLE
unclear about how to deal with the deprecation of IDENTITY for postgresql

### DISCUSSION THREAD

**Discussion by allan-simon** [doctrine/doctrine2#11038] [Support Questions] [answered]
I'm using doctrine/orm  2.16.2 

And I'm currently confused by the message  

```
Problem: Using the IDENTITY generator strategy with platform "Doctrine\DBAL\Platforms\PostgreSQL100Platform" is deprecated and will not be possible in Doctrine ORM 3.0.
Solution: Use the SEQUENCE generator strategy instead. (ClassMetadataFactory.php:638 called by ClassMetadataFactory.php:146, https://github.com/doctrine/orm/issues/8850, package doctrine/orm)
```

because:
   *  the recommended way for postgresql to generate `id` column is to use the sql standard `GENERATED BY DEFAULT AS IDENTITY`   
   * support for the above has been merged into `dbal`   https://github.com/doctrine/dbal/pull/5396/files  and will come with dbal 4.0 
   * the documentation also state  https://www.doctrine-project.org/projects/doctrine-orm/en/2.17/reference/basic-mapping.html#identifiers-primary-keys   "We now recommend using IDENTITY for PostgreSQL, " 

so why is it deprecated ? 

Link to the deprecation issue: https://github.com/doctrine/orm/issues/8850
Link to the PR that introduced the deprecation: https://github.com/doctrine/orm/pull/9833

**Comment by greg0ire** [comment]
I think something is indeed wrong especially since IDENTITY will become recommended again in https://github.com/doctrine/orm/pull/8931

Or maybe what's wrong is that your version of PG uses sequence emulated identity columns. It's hard for me to check because I'm on my phone, but if I were you I would look into the different PostgreSQLPlatform classes, and I would make sure I am using the latest version of Doctrine dbal.

EDIT: I think this will require DBAL 4: https://github.com/doctrine/dbal/pull/5396

**Reply by allan-simon** [reply]
I'm using postgresql 15 and so PostgreSQLPlatform100.  and yes sorry if it was not clear in my message, but  I understand that as of now as dbal4.0 is not yet released, there's no way to use `GENERATED ... AS IDENTITY` .

my main interrogation is that,  without dbal4.0  I would understand why it's marked deprecated because the current `IDENTITY` in postgresql  use `SERIAL` which is considered a bad practice by postgresql themselves https://wiki.postgresql.org/wiki/Don%27t_Do_This#Don.27t_use_serial.   But with dbal4.0 pointing its nose, it feels like we will do "deprecated/ removed/re-added" ? 

the only explanation I see is  "current IDENTITY is bad in postgresql (becaused backed by SERIAL under the hood) so people should migrate to something else (SEQUENCE ) waiting for the "good" identity to arrive ?

**Reply by greg0ire** [reply]
I think you got it right. I think you should probably not follow the deprecation because ultimately that means 2 migrations, and as it stands, the deprecation is only useful for new users in case it is easier to migrate from sequence than from serial. @morozov can you please chime in on this?

**Reply by greg0ire** [reply, marked as answer]
@allan-simon I think I've sorted this out: https://github.com/doctrine/orm/pull/11044

**Reply by allan-simon** [reply]
@greg0ire  ah great ! thanks a lot ,  with your PR things are much clearer

**Reply by morozov** [reply]
As it is implemented in https://github.com/doctrine/dbal/pull/5396, there is not upgrade path from serial to identity columns in DBAL 3.x.

What users most likely can do is migrate their database schema right before the deployment of DBAL 4. I.e.:

1. Deploy all pending application-owned schema changes.
2. Migrate the database schema using the script from the documentation.
3. Upgrade to DBAL 4.
4. Resume normal operation.

The only downside of this approach is that your application won't be able to make schema changes between steps 2 and 3 which seems negligible.

The feature is deprecated because it will be unavailable in the next major release. Usually, it also means that the users should update their code as recommended before the upgrade but it's not the case here.
